apps/labs/posts/close-encounter-with-pandas-and-the-jedis-of-open-source.md

```diff
@@ -0,0 +1,246 @@
+operations and tested them with empty inputs. It was really brilliant because it gave a very clear picture 
+of the various inconsistencies in handling empty inputs and allowed for interest in the issue to be 
+rekindled. Here's a cutout from the table showing some GroupBy methods acting on empty dataframes.
+
+<table>
+<thead>
+  <tr>
+    <th >Method</th>
+    <th >Example</th>
+    <th >Return</th>
+  </tr>
+</thead>
+<tbody>
+  <tr>
+  <td >apply()</td>
+  <td>
+
+  ```python
+  df = pd.DataFrame(data={'root_id': [], 'side': []})
+  df.groupby(['root_id']).apply(lambda x: x['side'].iloc[0])
+  ```
+    
+  </td>
+  <td>IndexError: single positional indexer is  out-of-bounds</td>
+  </tr>
+  <tr>
+  <td >take()</td>
+  <td >
+
+  ```python
+  df = pd.DataFrame({'a': [], 'b': []})
+  df.groupby('a').take([0])
+  ```
+
+  </td>
+  <td >IndexError: single positional indexer is  out-of-bounds</td>
+  </tr>
+  <tr>
+  <td >sample()</td>
+  <td >
+
+  ```python
+  df = pd.DataFrame({'a': [], 'b': []})
+  df.groupby('a').sample()
+  ```
+
+  </td>
+  <td >ValueError: need at least one array to concatenate</td>
+  </tr>
+  <tr>
+    <td >boxplot()</td>
+    <td >
+
+  ```python
+  df.groupby('a').boxplot()
+  ```
+
+  </td>
+  <td >ValueError: Number of columns must be a positive integer, not 0</td>
+  </tr>
+</tbody>
+</table>
+
+### Gaining an audience with pandas' Jedi Council
+
+Communicating these findings was the next challenge. There wasn't a clear 
+understanding for me of what consistency needed to be maintained. I 
+created a few issues, made some pull requests, and got very few responses 
+and comments. More discussion was needed. Noa, being an excellent 
+communicator, secured an audience with some pandas maintainers and after a 
+few comments from maintainers like [Joris Van den Bossche](https://github.com/jorisvandenbossche), 
+[Richard Shadrach](https://github.com/rhshadrach) and [JHM Darbyshire](https://github.com/attack68), 
+a better picture on what is expected and the best approach to fixing the issue was painted.
+
+### Debugging, ain't like dusting crops
+
+One of the most exciting parts of my internship has been learning how experienced developers went about 
+debugging and solving problems. Jaime held a live-coding section where he did exactly that with a project 
+he was working on for conda forge. It was an issue with errors on the CI (continuous integration) that 
+wasn't reproducible on his local machine. He had a list of hypotheses and he meticulously updated this 
+list as new information was presented from each fix he attempted. It was obvious that "*the force is strong 
+with this one*" and I had so much fun learning and laughing at both the situation and his jokes (he's a 
+very funny jedi 😆). 
+
+Aaron had a very particular way of seeing a problem and accessing a possible solution. It involved looking 
+at how other packages handled similar problems, trying out examples (using toys as Noa once mentioned), 
```

Comment:
Hmm, I may have used the term [Toy Problem](https://en.wikipedia.org/wiki/Toy_problem) to denote a simple (or simplified) use case? Is that what you meant here, or did I refer to actual toys?!

Reply:
I think it was more of a Toy Program. Just like small scripts to test out an idea. I can totally change that to a simpler term (maybe I was projecting, although I have actually not heard the term used before).